File: frontend/src/types/poi.ts
// Mirrors backend/models/poi.py — keep in sync when fields change

export type POIType = "settlement" | "encampment" | "dungeon" | "ruins";

export interface POI {
  id: string;
  adventure_id: string;
  map_id: string;
  tile_x: number;
  tile_y: number;
  type: POIType;
  tier: 1 | 2 | 3;
  detail_card_id: string | null;
  generated: boolean;
}

export interface POIUpdate {
  detail_card_id?: string | null;
  generated?: boolean;
}

// ── Dungeon ───────────────────────────────────────────────────────────────────

export type ExitDirection = "north" | "south" | "east" | "west" | "down" | "up";

export interface Exit {
  direction: ExitDirection;
  leads_to_room_id: string | null;
  leads_to_floor: number | null;
}

export interface Dungeon {
  id: string;
  adventure_id: string;
  poi_id: string | null;
  ruin_structure_id: string | null;
  floor_count: number;
  door_budget_per_floor: Record<string, number>;
  stairs_placed: Record<string, boolean>;
  floors_initialized: number[];
}

export interface DungeonRoom {
  id: string;
  adventure_id: string;
  dungeon_id: string;
  floor: number;
  x: number;
  y: number;
  is_entrance: boolean;
  is_boss_room: boolean;
  exits: Exit[];
  content: string[];
}

export interface DungeonRoomUpdate {
  exits?: Exit[];
  content?: string[];
}

// ── Settlement ────────────────────────────────────────────────────────────────

export interface SettlementLocation {
  id: string;
  name: string | null;
  detail_card_id: string | null;
}

export interface Settlement {
  id: string;
  adventure_id: string;
  poi_id: string;
  location_count: number;
  locations: SettlementLocation[];
  generated: boolean;
}

export interface SettlementUpdate {
  locations?: SettlementLocation[];
  generated?: boolean;
}

// ── Ruins ─────────────────────────────────────────────────────────────────────

export interface RuinStructure {
  id: string;
  label: string | null;
  floor_count: number;
  dungeon_id: string | null;
}

export interface Ruin {
  id: string;
  adventure_id: string;
  poi_id: string;
  structures: RuinStructure[];
  generated: boolean;
}

// ── Request bodies ────────────────────────────────────────────────────────────

export interface DiscoverRequest {
  adventure_id: string;
  map_id: string;
  tile_x: number;
  tile_y: number;
}

export interface EnterRequest {
  adventure_id: string;
  poi_id: string;
}

export interface ExploreRequest {
  from_room_id: string;
  direction: ExitDirection;
}

export interface SettlementEnterRequest {
  adventure_id: string;
  poi_id: string;
}

export interface RuinEnterRequest {
  adventure_id: string;
  poi_id: string;
}

File: backend/models/poi.py
from typing import Literal
from pydantic import BaseModel, Field
from .shared import BaseDocument, new_id


POIType = Literal["settlement", "encampment", "dungeon", "ruins"]


# ── POI ───────────────────────────────────────────────────────────────────────

class POICreate(BaseModel):
    adventure_id: str
    map_id: str
    tile_x: int
    tile_y: int
    type: POIType
    tier: int
    detail_card_id: str | None = None


class POI(BaseDocument):
    map_id: str
    tile_x: int
    tile_y: int
    type: POIType
    tier: int
    detail_card_id: str | None = None
    generated: bool = False


class POIUpdate(BaseModel):
    detail_card_id: str | None = None
    generated: bool | None = None


# ── Dungeon ───────────────────────────────────────────────────────────────────

class Exit(BaseModel):
    direction: Literal["north", "south", "east", "west", "down", "up"]
    leads_to_room_id: str | None = None   # None = promise
    leads_to_floor: int | None = None     # stair exits only


class Dungeon(BaseDocument):
    poi_id: str | None = None             # None if this dungeon belongs to a ruin structure
    ruin_structure_id: str | None = None  # None if standalone dungeon POI
    floor_count: int
    door_budget_per_floor: dict[str, int]
    stairs_placed: dict[str, bool] = Field(default_factory=dict)
    floors_initialized: list[int] = Field(default_factory=list)


class DungeonRoom(BaseModel):
    id: str = Field(default_factory=new_id)
    adventure_id: str
    dungeon_id: str
    floor: int
    x: int
    y: int
    is_entrance: bool = False
    is_boss_room: bool = False
    exits: list[Exit] = Field(default_factory=list)
    content: list[str] = Field(default_factory=list)


class DungeonRoomUpdate(BaseModel):
    exits: list[Exit] | None = None
    content: list[str] | None = None


# ── Settlement ────────────────────────────────────────────────────────────────

class SettlementLocation(BaseModel):
    id: str = Field(default_factory=new_id)
    name: str | None = None
    detail_card_id: str | None = None


class Settlement(BaseDocument):
    poi_id: str
    location_count: int
    locations: list[SettlementLocation] = Field(default_factory=list)
    generated: bool = False


class SettlementUpdate(BaseModel):
    locations: list[SettlementLocation] | None = None
    generated: bool | None = None


# ── Ruins ─────────────────────────────────────────────────────────────────────

class RuinStructure(BaseModel):
    id: str = Field(default_factory=new_id)
    label: str | None = None
    floor_count: int
    dungeon_id: str | None = None


class Ruin(BaseDocument):
    poi_id: str
    structures: list[RuinStructure] = Field(default_factory=list)
    generated: bool = False


class RuinUpdate(BaseModel):
    structures: list[RuinStructure] | None = None


# ── Request bodies ────────────────────────────────────────────────────────────

class DiscoverRequest(BaseModel):
    adventure_id: str
    map_id: str
    tile_x: int
    tile_y: int


class EnterRequest(BaseModel):
    adventure_id: str
    poi_id: str


class ExploreRequest(BaseModel):
    from_room_id: str
    direction: str


class SettlementEnterRequest(BaseModel):
    adventure_id: str
    poi_id: str


class RuinEnterRequest(BaseModel):
    adventure_id: str
    poi_id: str
